### PR TITLE
[chore][golden] Normal timestamps in golden metric files

### PR DESCRIPTION
**Pull request by BominRahmani**
**Description:** Normalized the timestamps on Golden file outputs.
Normalization in this instance basically means setting non-unique timestamps of a metric equal to each other and unique timestamps incremented in increments of 1,000,000 nanoseconds. To see a clear example of this, navigate into the added test files (internal/coreinternal/golden/testdata/timestamp-norm)

The goal is to further make the Golden output files as deterministic as possible and clear up the cluttered git diff's every time one of these golden files are remade.

**Testing:** Test file testing normalization added within the golden package files.

**Documentation:**

### CODE REVIEW

File: internal/coreinternal/golden/normalize_timestamps.go [hunk 1]
```diff
@@ -0,0 +1,250 @@
...
+					for l := 0; l < m.Summary().DataPoints().Len(); l++ {
+						m.Summary().DataPoints().At(l).SetStartTimestamp(nanosec)
+					}
+				}
+			}
+		}
+	}
+}
+
+func sortMetricDataPointSlices(ms pmetric.Metrics) {
```
Comment (djaglowski) [inline internal/coreinternal/golden/normalize_timestamps.go:73]: This would be helpful regardless of how we normalize timestamps. Would you mind peeling this off into another PR and regenerating files accordingly?

File: internal/coreinternal/golden/normalize_timestamps.go [hunk 2]
```diff
@@ -0,0 +1,250 @@
...
+
+			}
+		}
+	}
+}
+
+func sortNumberDataPointSlice(ndps pmetric.NumberDataPointSlice) {
+	ndps.Sort(func(a, b pmetric.NumberDataPoint) bool {
+		aTime := a.Timestamp().String()
+		bTime := b.Timestamp().String()
```
Comment (djaglowski) [inline internal/coreinternal/golden/normalize_timestamps.go:99]: I think we need to incorporate more information in order to make this a deterministic sort. Specifically, we need to include attributes in the sort order.

For example, the timestamps on these data point could change and it would cause us to sort differently:
```
datapoints:
  - attributes:
    -  foo: bar
    timestamp: 1000
  - attributes:
        hello: world
    timestamp: 1001 # if changed to 999, this would sort to first position
```

Instead, we should sort first based on attributes, then on timestamps.

File: internal/coreinternal/golden/normalize_timestamps.go [hunk 3]
```diff
@@ -0,0 +1,167 @@
...
+}
+
+func compareTimestamps(t1, t2 pcommon.Timestamp) int {
+	if t1 < t2 {
+		return -1
+	} else if t1 > t2 {
+		return 1
+	}
+	return 0
+}
```
Comment (djaglowski) [inline internal/coreinternal/golden/normalize_timestamps.go:167]: This is only used in one place so can just be inline in the sort call.

File: internal/coreinternal/golden/golden.go [hunk 4]
```diff
@@ -70,6 +70,7 @@ func MarshalMetricsYAML(metrics pmetric.Metrics) ([]byte, error) {
 // writeMetrics writes a pmetric.Metrics to the specified file in YAML format.
 func writeMetrics(filePath string, metrics pmetric.Metrics) error {
 	sortMetrics(metrics)
+	normalizeTimestamp(metrics)
```
Comment (djaglowski) [inline internal/coreinternal/golden/golden.go:73]:
```suggestion
	normalizeTimestamps(metrics)
```

File: internal/coreinternal/golden/normalize_timestamps.go [hunk 5]
```diff
@@ -0,0 +1,167 @@
...
+					normExponentialHistogramDataPointSlice(m.ExponentialHistogram().DataPoints())
+				case pmetric.MetricTypeSummary:
+					normSummaryDataPointSlice(m.Summary().DataPoints())
+				}
+			}
+		}
+	}
+}
+
+func zeroStartTimestamp(metrics pmetric.Metrics) {
```
[the remark this answers is not in the document]
Comment (BominRahmani) [reply internal/coreinternal/golden/normalize_timestamps.go:38]: Not specifically. I started out trying to make it a bit flexible, so that if we wanted we could apply an option whether to zero the startTimestamps or not. But looking back it's probably fine to just combine these two methods into one.

**Comment by MovieStoreGuy** [comment]
Would it be possible to use a virtual clock instead of the approach taken?

The reason is that you can simplify the developer experience with something like `"github.com/benbjohnson/clock"` or `"github.com/tilinna/clock"` to allow for porting across different testing patterns.

**Comment by djaglowski** [comment]
@MovieStoreGuy, do you mind elaborating on how a virtual clock could be used here? The idea with this PR is to stabilize the golden files so that regenerating them doesn't cause huge diffs. It's not clear to me how a virtual clock would help with this.

File: internal/coreinternal/golden/testdata/timestamp-norm/after_normalize.yaml [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+                  timeUnixNano: "1000000"
+                - asInt: "5"
+                  startTimeUnixNano: "0"
+                  timeUnixNano: "1000000"
+                - asInt: "10"
+                  startTimeUnixNano: "0"
+                  timeUnixNano: "500000"
+                - asInt: "50"
+                  startTimeUnixNano: "0"
+                  timeUnixNano: "2000000"
```
Comment (djaglowski) [inline internal/coreinternal/golden/testdata/timestamp-norm/after_normalize.yaml:73]: Do we have any tests that apply both sorting and timestamp normalization? I can appreciate test them in isolation, but ultimately we want to see the result of both as well.

File: receiver/apachereceiver/testdata/scraper/expected.yaml [hunk 7]
```diff
@@ -47,7 +47,7 @@ resourceMetrics:
                     - key: mode
                       value:
                         stringValue: system
-                  timeUnixNano: "1632495518500962000"
+                  timeUnixNano: "500000"
```
Comment (djaglowski) [inline receiver/apachereceiver/testdata/scraper/expected.yaml:50]: 500,000 seems pretty arbitrary. If we 2x the values it will be easier to visually count:
- `1000000`
- `2000000`

File: internal/coreinternal/golden/testdata/timestamp-norm/after_normalize.yaml (hunk 6 above)
Comment (BominRahmani) [reply internal/coreinternal/golden/testdata/timestamp-norm/after_normalize.yaml:73]: Wouldn't the results of both end up being any of the current golden files in the change log since the sort work has been applied to the golden files and pushed in a previous PR, while the timestamps are being applied here? I can definitely make another test if you still would like one, but seeing as the timestamp normalization works independent of any sorting, the test would be purely for visual purposes.

Comment (djaglowski) [reply internal/coreinternal/golden/testdata/timestamp-norm/after_normalize.yaml:73]: The purpose would be as an overall test of the golden package. The actual golden files would use the functionality, but the testing of the golden package should be done locally.

File: internal/coreinternal/golden/testdata/sort-metrics/after.yaml [hunk 8]
```diff
@@ -157,4 +157,4 @@ resourceMetrics:
               value:
                 stringValue: AAAA
           name: otelcol/questionablyfakereceiver
-          version: latest
+          version: latest
```
Comment (djaglowski) [inline internal/coreinternal/golden/testdata/sort-metrics/after.yaml:160]:
```suggestion
          version: latest

```

File: internal/coreinternal/golden/testdata/standardize-metrics/after.yaml [hunk 9]
```diff
@@ -0,0 +1,183 @@
...
+                  timeUnixNano: "1000000"
+              isMonotonic: true
+            unit: By
+        scope:
+          attributes:
+            - key: b
+              value:
+                stringValue: BBB
+          name: otelcol/standardization
+          version: latest
```
Comment (djaglowski) [inline internal/coreinternal/golden/testdata/standardize-metrics/after.yaml:183]:
```suggestion
          version: latest

```

File: internal/coreinternal/golden/testdata/standardize-metrics/before.yaml [hunk 10]
```diff
@@ -0,0 +1,205 @@
...
+                  timeUnixNano: "8888888888888888888"
+              isMonotonic: false
+            unit: "1"
+        scope:
+          name: otelcol/FakeMetric
+          version: latest
+          attributes:
+            - key: a
+              value:
+                stringValue: AAA
```
Comment (djaglowski) [inline internal/coreinternal/golden/testdata/standardize-metrics/before.yaml:205]:
```suggestion
                stringValue: AAA

```

File: internal/coreinternal/golden/testdata/timestamp-norm/after_normalize.yaml [hunk 11]
```diff
@@ -0,0 +1,78 @@
...
+                  startTimeUnixNano: "0"
+                  timeUnixNano: "1000000"
+                - asInt: "50"
+                  startTimeUnixNano: "0"
+                  timeUnixNano: "4000000"
+              isMonotonic: true
+            unit: By
+        scope:
+          name: otelcol/timestampNormalization
+          version: latest
```
Comment (djaglowski) [inline internal/coreinternal/golden/testdata/timestamp-norm/after_normalize.yaml:78]:
```suggestion
          version: latest

```

File: internal/coreinternal/golden/testdata/timestamp-norm/before_normalize.yaml [hunk 12]
```diff
@@ -0,0 +1,78 @@
...
+                  startTimeUnixNano: "1924857624581089658"
+                  timeUnixNano: "1111111111111111111"
+                - asInt: "50"
+                  startTimeUnixNano: "1924857624581089658"
+                  timeUnixNano: "8888888888888888888"
+              isMonotonic: true
+            unit: By
+        scope:
+          name: otelcol/timestampNormalization
+          version: latest
```
Comment (djaglowski) [inline internal/coreinternal/golden/testdata/timestamp-norm/before_normalize.yaml:78]:
```suggestion
          version: latest

```

File: internal/coreinternal/golden/testdata/standardize-metrics/after.yaml [hunk 13]
```diff
@@ -0,0 +1,183 @@
...
+                    - key: a
+                      value:
+                        stringValue: AAAA
+                  timeUnixNano: "1000000"
+                - asInt: "0"
+                  attributes:
+                    - key: a
+                      value:
+                        stringValue: AAAA
+                  timeUnixNano: "1000000"
```
Comment (djaglowski) [inline internal/coreinternal/golden/testdata/standardize-metrics/after.yaml:69]: Something appears to be wrong here because we have two data points with the same attributes and timestamps.

File: internal/coreinternal/golden/testdata/standardize-metrics/after.yaml [hunk 14]
```diff
@@ -0,0 +1,183 @@
...
+                - asInt: "30"
+                  attributes:
+                    - key: a
+                      value:
+                        stringValue: AAAA
+                  timeUnixNano: "1000000"
```
Comment (djaglowski) [inline internal/coreinternal/golden/testdata/standardize-metrics/after.yaml:31]: Here too

File: internal/coreinternal/golden/normalize_timestamps_test.go [hunk 15]
```diff
@@ -0,0 +1,22 @@
...
+package golden // import "github.com/open-telemetry/opentelemetry-collector-contrib/internal/coreinternal/golden"
+
+import (
+	"path/filepath"
+	"testing"
+
+	"github.com/stretchr/testify/require"
+)
+
+func TestNormalize(t *testing.T) {
```
Comment (djaglowski) [inline internal/coreinternal/golden/normalize_timestamps_test.go:13]:
```suggestion
func TestNormalizeTimestamps(t *testing.T) {
```

File: internal/coreinternal/golden/normalize_timestamps.go [hunk 16]
```diff
@@ -0,0 +1,198 @@
...
+						m.Summary().DataPoints().At(l).SetStartTimestamp(nanosec)
+					}
+					normSummaryDataPointSlice(m.Summary().DataPoints())
+				}
+			}
+		}
+	}
+}
+
+func normGaugeDataPointSlice(ndps pmetric.NumberDataPointSlice) {
```
Comment (djaglowski) [inline internal/coreinternal/golden/normalize_timestamps.go:56]: Since this is used for gauge and sum
```suggestion
func normNumberDataPointSlice(ndps pmetric.NumberDataPointSlice) {
```

File: internal/coreinternal/golden/normalize_timestamps.go [hunk 17]
```diff
@@ -0,0 +1,198 @@
...
+	hashCache := make(map[[16]byte]bool)
```
Comment (djaglowski) [inline internal/coreinternal/golden/normalize_timestamps.go:57]: What are we keeping track of here? The name should reflect the meaning.

File: internal/coreinternal/golden/normalize_timestamps.go [hunk 18]
```diff
@@ -0,0 +1,198 @@
...
+	for i := 0; i < ndps.Len(); i++ {
+		attrHash := pdatautil.MapHash(ndps.At(i).Attributes())
+		if hashCache[attrHash] {
+			continue
+		}
+		ts := []pcommon.Timestamp{ndps.At(i).Timestamp()}
```
Comment (djaglowski) [inline internal/coreinternal/golden/normalize_timestamps.go:63]: Let's call this `timeSeries`, since datapoints with the same attributes make up a time series.

File: internal/coreinternal/golden/normalize_timestamps.go [hunk 19]
```diff
@@ -0,0 +1,198 @@
...
+		for j := i + 1; j < ndps.Len(); j++ {
```
Comment (djaglowski) [inline internal/coreinternal/golden/normalize_timestamps.go:64]:
```suggestion
		// Find any other data points in the time series
		for j := i + 1; j < ndps.Len(); j++ {
```

File: internal/coreinternal/golden/normalize_timestamps.go [hunk 20]
```diff
@@ -0,0 +1,198 @@
...
+			if attrHash == pdatautil.MapHash(ndps.At(j).Attributes()) {
+				ts = append(ts, ndps.At(j).Timestamp())
+			}
```
Comment (djaglowski) [inline internal/coreinternal/golden/normalize_timestamps.go:67]:
```suggestion
			if pdatautil.MapHash(ndps.At(j).Attributes()) != attrHash{
				continue
			}	
			ts = append(ts, ndps.At(j).Timestamp())
```

It's idiomatic in golang to keep the code flatter by continuing/returning early.

File: internal/coreinternal/golden/normalize_timestamps.go [hunk 21]
```diff
@@ -0,0 +1,198 @@
...
+		}
+
+		normalizedTs := normalizeTimestampSlice(ts)
+		for k := 0; k < ndps.Len(); k++ {
+			if pdatautil.MapHash(ndps.At(k).Attributes()) == attrHash {
```
Comment (djaglowski) [inline internal/coreinternal/golden/normalize_timestamps.go:72]: Same here, invert condition and continue, rather than nest conditionals

File: internal/coreinternal/golden/normalize_timestamps.go [hunk 22]
```diff
@@ -0,0 +1,198 @@
...
+				}
+			}
+		}
+		hashCache[attrHash] = true
+	}
+}
+
+// sorts a slice of pcommon.Timestamp and returns a normalized slice of pcommon.Timestamp
+// normalization entails setting nonunique subsequent timestamps to the same value while incrementing unique timestamps by a set value of 1,000,000 ns
+func normalizeTimestampSlice(ts []pcommon.Timestamp) map[pcommon.Timestamp]pcommon.Timestamp {
```
Comment (djaglowski) [inline internal/coreinternal/golden/normalize_timestamps.go:170]:
```suggestion
func normalizeTimeSeries(ts []pcommon.Timestamp) map[pcommon.Timestamp]pcommon.Timestamp {
```

Comment (djaglowski) [inline internal/coreinternal/golden/normalize_timestamps.go:170]: Why do we need to return anything here? Aren't we setting the timestamps?

File: internal/coreinternal/golden/normalize_timestamps.go [hunk 23]
```diff
@@ -0,0 +1,198 @@
...
+	normalizedTs := make(map[pcommon.Timestamp]pcommon.Timestamp)
+	tempTs := make([]pcommon.Timestamp, len(ts))
```
Comment (djaglowski) [inline internal/coreinternal/golden/normalize_timestamps.go:172]: Why do we need a temp/copy?

File: internal/coreinternal/golden/normalize_timestamps.go [hunk 24]
```diff
@@ -0,0 +1,198 @@
...
+			return 0
+		}(tempTs[i], tempTs[j]) < 0
+	})
+
+	rc := time.Nanosecond * 0
+	prevTime := pcommon.NewTimestampFromTime(time.Unix(0, 0))
+
+	for i := range ts {
+		if tempTs[i] != prevTime {
+			rc += 1000000 * time.Nanosecond
```
Comment (djaglowski) [inline internal/coreinternal/golden/normalize_timestamps.go:190]: Let's move this into a function so we can isolate the logic and change it later if necessary. It shouldn't be dependent on knowing the previous data point's time - just make it functional. Roughly: 
```go
func normalTime(timeSeriesIndex int) pcommon.Timestamp {
    return pcommon.NewTimestampFromTime(time.Unix(0, 0).Add(i * time.Second))
}
```

File: internal/coreinternal/golden/normalize_timestamps.go [hunk 25]
```diff
@@ -0,0 +1,198 @@
...
+	}
+}
+
+// sorts a slice of pcommon.Timestamp and returns a normalized slice of pcommon.Timestamp
+// normalization entails setting nonunique subsequent timestamps to the same value while incrementing unique timestamps by a set value of 1,000,000 ns
+func normalizeTimestampSlice(ts []pcommon.Timestamp) map[pcommon.Timestamp]pcommon.Timestamp {
+	normalizedTs := make(map[pcommon.Timestamp]pcommon.Timestamp)
+	tempTs := make([]pcommon.Timestamp, len(ts))
+	copy(tempTs, ts)
+	sort.Slice(tempTs, func(i, j int) bool {
```
Comment (djaglowski) [inline internal/coreinternal/golden/normalize_timestamps.go:174]: I don't think we need the temp slice. We created the timeseries slice and can just sort it, then set the timestamps according to the resulting index. It won't affect the order of the _data point_ slice.

File: internal/coreinternal/golden/normalize_timestamps.go [hunk 26]
```diff
@@ -0,0 +1,198 @@
...
+		}
+		ts := []pcommon.Timestamp{ndps.At(i).Timestamp()}
+		for j := i + 1; j < ndps.Len(); j++ {
+			if attrHash == pdatautil.MapHash(ndps.At(j).Attributes()) {
+				ts = append(ts, ndps.At(j).Timestamp())
+			}
+		}
+
+		normalizedTs := normalizeTimestampSlice(ts)
+		for k := 0; k < ndps.Len(); k++ {
```
Comment (djaglowski) [inline internal/coreinternal/golden/normalize_timestamps.go:71]: Why do we need this loop? Seems redundant with `normalizeTimeSeries`

File: internal/coreinternal/golden/testdata/standardize-metrics/after.yaml (hunk 13 above)
Comment (BominRahmani) [reply internal/coreinternal/golden/testdata/standardize-metrics/after.yaml:69]: This is because the original for those
image
are both the same timeUnixNano. I was under the understanding that in order for the timestamps to scale up, they must
1. have the same attribute
2. have timestamps that differ from each other.
Would it never be the case that in a time series where the attributes are the same for a metric, the times stamps are also the same. Or in other words would you ever have a situation like the picture above, where within a metric you have datapoints of the same attributes and same timestamp?

Comment (djaglowski) [reply internal/coreinternal/golden/testdata/standardize-metrics/after.yaml:69]: > would you ever have a situation like the picture above, where within a metric you have datapoints of the same attributes and same timestamp?

I don't believe this is valid, for sums and gauges at least. It's not strictly prohibited, but I think we don't need to handle it here.

**Comment by MovieStoreGuy** [comment]
Considering that you can ignore the timestamps with https://github.com/open-telemetry/opentelemetry-collector-contrib/pkg/pdatatest, is the value still here for this PR?

cc: @djaglowski @BominRahmani ?

**Comment by djaglowski** [comment]
@MovieStoreGuy, the option to ignore timestamps is only useful when comparing them. The purpose of this PR would be to stabilize the yaml files that we commit to the repo. Currently, these timestamps are often regenerated and result in large diffs that unnecessarily bury the meaningful changes.

File: internal/coreinternal/golden/normalize_timestamps.go (hunk 22 above)
Comment (BominRahmani) [reply internal/coreinternal/golden/normalize_timestamps.go]: This is a bit of a workaround solution, instead of setting the timestamps by passing in the actual DataPointSlice, I am abstracting out the logic to take in just a timestamp slice, that way i can sort and normalize here then reassign those values in the normDataPointSlice methods. It saves me out on having multiple NormalizeTimeSeries methods for each DataPoint type.

File: internal/coreinternal/golden/normalize_timestamps.go (hunk 26 above)
Comment (BominRahmani) [reply internal/coreinternal/golden/normalize_timestamps.go:74]: NormalizeTimeSeries does not assign values to the DatapointSlice, it sorts and normalizes a slice of timestamps which are reassigned here.

File: receiver/azuremonitorreceiver/testdata/expected_metrics/metrics_golden.yaml [hunk 27]
```diff
@@ -215,8 +201,7 @@ resourceMetrics:
                     - key: azuremonitor.resource_id
                       value:
                         stringValue: resourceId2
-                  startTimeUnixNano: "1681790471654481953"
```
Comment (djaglowski) [inline receiver/azuremonitorreceiver/testdata/expected_metrics/metrics_golden.yaml:218]: Why are we losing the start time?

Comment (BominRahmani) [reply receiver/azuremonitorreceiver/testdata/expected_metrics/metrics_golden.yaml:218]: Since startTimeUnixNano is already a non required field, i set them all to 0, doing this gets rid of them. I figured since we are looking to not only reduce git clutter, but also make golden files a bit more readable, it would make sense to only keep the timeUnixNano and normalize those values. Is there a reason we would want to keep them and normalize those values also?

File: internal/coreinternal/golden/normalize_timestamps.go [hunk 28]
```diff
@@ -0,0 +1,151 @@
...
+				}
+			}
+		}
+	}
+}
+
+type timeStampInfo struct {
+	timestamp pcommon.Timestamp
+	position  int
+}
```
Comment (djaglowski) [inline internal/coreinternal/golden/normalize_timestamps.go:42]: I don't think we need this struct, or any of the position tracking.

All we need is:
- Sort a `[]pcommon.Timestamp` that contains all `Timestamp` and `StartTimestamp`, regardless of uniqueness
- Create `normalizedTs` the same as you are doing
- Iterate over all DataPoints, and use `normalizeTs` to determine the appropriate value to set each timestamp
```go
for _, dp := range dps {
    dp.SetTimestamp(normalizeTs[dp.Timestamp()])
    dp.SetStartTimestamp(normalizeTs[dp.StartTimestamp()])
}
```

File: exporter/signalfxexporter/go.mod [hunk 29]
```diff
@@ -32,6 +32,7 @@ require (
 )
 
 require (
+	github.com/cespare/xxhash/v2 v2.2.0 // indirect
```
Comment (djaglowski) [inline exporter/signalfxexporter/go.mod:35]: Changes to this file look unrelated

**Review by djaglowski — APPROVED** [review]

File: internal/coreinternal/golden/normalize_timestamps.go [hunk 30]
```diff
@@ -0,0 +1,117 @@
...
+			continue
+		}
+		if _, ok := normalizedTs[v]; !ok {
+			normalizedTs[v] = normalTime(count)
+			count++
+		}
+	}
+	for i := range timeSeries {
+		timeSeries[i] = normalizedTs[timeSeries[i]]
+	}
```
Comment (djaglowski) [inline internal/coreinternal/golden/normalize_timestamps.go:67]: What's the point of this if we're just going to return normalizedTs?
